### PR TITLE
Maintainer's checklist on accepting formula PR

### PR DESCRIPTION
**Pull request by igas**
On the slack topic of maintainer checklist came up and this is the first draft of it.

I'm not sure at this stage where to link it and how much it fits in the docs (there are many overlaps with formula acceptance, formula cookbook and maintainer guidelines).

Also, the introduction was not very thought through and feedback there much appreciated.

At this stage, I tried to be more focused on the completeness of the list.

Thanks @fxcoudert for most of the items.

I know @SMillerDev and @Moisan have many snippets and replies that can benefit this guide too.

### CODE REVIEW

**Comment by SMillerDev** [comment]
I think the replies could just be added to the organization?

**Comment by igas** [comment]
@SMillerDev this is a good idea! I think it's worth mentioning them in some way to bring awareness to them.

**Comment by issyl0** [comment]
Should we add the checklist for the Linux side of things here (I'm happy to write one)? Or have a separate page of checklist for that? (On that, there's also [`Linuxbrew/homebrew-developer` docs](https://github.com/Linuxbrew/homebrew-developer/tree/master/docs) which doesn't have anything about workflows, they're just passed on by word of mouth.) Things are slightly different on the `brew pull` side and on the workflow side particularly where bottles are involved.

**Comment by SMillerDev** [comment]
I think the linux side should be documented on equal footing with the macOS flow. So maybe just suggest adjustments here?

File: docs/Formula-Merge-Checklist.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+burnout.
+
+This is a more practical checklist, it should be used after you get familiar with
+[Maintainer Guidelines](Maintainer-Guidelines.md).
+
+## Checklist
+
+Check for:
+
+- previously opened active PRs, we would like to be fair to contributors who came first;
```
Comment (MikeMcQuaid) [inline docs/Formula-Merge-Checklist.md]: Can skip the trailing `;` on each line I think

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+This is a more practical checklist, it should be used after you get familiar with
+[Maintainer Guidelines](Maintainer-Guidelines.md).
+
+## Checklist
+
+Check for:
+
+- previously opened active PRs, we would like to be fair to contributors who came first
+- patches/`inreplace` that been applied to upstream and can be removed
```
Comment (javian) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: that *have* been

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+- comments in formula around `url`, we do skip some version (for example [vim](https://github.com/Homebrew/homebrew-core/blob/359dbb190bb3776c4d6a1f603a271dd8f186f077/Formula/vim.rb#L4) or [v8](https://github.com/Homebrew/homebrew-core/blob/359dbb190bb3776c4d6a1f603a271dd8f186f077/Formula/v8.rb#L4))
```
Comment (javian) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: /version/versions/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+- vendored resources that need updates (for example [emscripten](https://github.com/Homebrew/homebrew-core/commit/57126ac765c3ac5201ce53bcdebf7a0e19071eba))
+- vendored dependencies (for example [certbot](https://github.com/Homebrew/homebrew-core/pull/42966/files))
+- stable/announced release
+  - some teams use odd minor release number for tests and even for stable releases
+  - other teams drop new version with minor release 0 but promote it to stable only after few minor releases
+  - if the software uses only hosted version control (such as github, gitlab or bitbucket), the release should be tagged and if upstream marks latest/pre-releases, PR must use latest
+- does changelog mention addition/removal of dependency and it's addressed in the PR
+  - does formula depend on versioned formula (for example `python@2`, `go@1.10`, `erlang@17`) that can be upgraded
+- commits
+  - contain one formula change per commit
```
Comment (javian) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: I think this can be easily misinterpreted and we could add a line here on how to handle multiple commits like
* Ask author to squash
* rebase during merge

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  - version update follows preferred message format for simple version updates `foobar 7.3`
+  - other fixes format is `foobar: fix flibble matrix`
+  - you can use `--bump` flag for `brew pull` in case PR have single commit but the wrong message
+- bottle block is not removed
+  Suggested reply:
+  ```
+  Please keep bottle block in place, [@BrewTestBot](https://github.com/BrewTestBot) takes care of it.
+  ```
+- is there are test block for other than checking version or printing help? Consider asking to add one
```
Comment (scpeters) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]:
```suggestion
- is there a test block for other than checking version or printing help? Consider asking to add one
```

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
+# Homebrew/homebrew-core Merge Checklist
+
+The following checklist is intended to help maintainers to decide on
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: remove second "to"

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+whether to merge, request changes or close a PR. It also brings more
+transparency for contributors in addition to
+[Acceptable Formulae](Acceptable-Formulae.md) requirements.
+
+This is a guiding principle. As a maintainer, you can make a call on either
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: /either/whether to/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+request changes from a contributor or help them out based on their comfort and
+previous contributions. Remember, as a team we
+[Prioritise Maintainers Over Users](Maintainers-Avoiding-Burnout.md) to avoid
+burnout.
+
+This is a more practical checklist, it should be used after you get familiar with
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: /,/;/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+[Maintainer Guidelines](Maintainer-Guidelines.md).
+
+## Checklist
+
+Check for:
+
+- previously opened active PRs, we would like to be fair to contributors who came first
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: /, we/, as we/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md (hunk 3 above)
Comment (EricFromCanada) [reply docs/Homebrew-homebrew-core-Merge-Checklist.md]: /, we/, as we/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+- patches/`inreplace` that been applied to upstream and can be removed
+- comments in formula around `url`, we do skip some version (for example [vim](https://github.com/Homebrew/homebrew-core/blob/359dbb190bb3776c4d6a1f603a271dd8f186f077/Formula/vim.rb#L4) or [v8](https://github.com/Homebrew/homebrew-core/blob/359dbb190bb3776c4d6a1f603a271dd8f186f077/Formula/v8.rb#L4))
+- vendored resources that need updates (for example [emscripten](https://github.com/Homebrew/homebrew-core/commit/57126ac765c3ac5201ce53bcdebf7a0e19071eba))
+- vendored dependencies (for example [certbot](https://github.com/Homebrew/homebrew-core/pull/42966/files))
+- stable/announced release
+  - some teams use odd minor release number for tests and even for stable releases
+  - other teams drop new version with minor release 0 but promote it to stable only after few minor releases
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: /after few/after a few/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  - if the software uses only hosted version control (such as github, gitlab or bitbucket), the release should be tagged and if upstream marks latest/pre-releases, PR must use latest
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: GitHub, GitLab or Bitbucket

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+- does changelog mention addition/removal of dependency and it's addressed in the PR
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: /it's/is it/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  - does formula depend on versioned formula (for example `python@2`, `go@1.10`, `erlang@17`) that can be upgraded
+- commits
+  - contain one formula change per commit
+  - version update follows preferred message format for simple version updates `foobar 7.3`
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: add colon after "updates"

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  - other fixes format is `foobar: fix flibble matrix`
+  - you can use `--bump` flag for `brew pull` in case PR have single commit but the wrong message
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: /PR have/the PR has a/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+- bottle block is not removed
+  Suggested reply:
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: There needs to be a blank line before this to create a new paragraph within the list item

File: docs/Homebrew-homebrew-core-Merge-Checklist.md (hunk 5 above)
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: /are/a/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+  ```
+  Please keep bottle block in place, [@BrewTestBot](https://github.com/BrewTestBot) takes care of it.
+  ```
+- is there are test block for other than checking version or printing help? Consider asking to add one
+- if CI failed
+  - due to test block - paste relevant lines and add `test failure` label
+  - due to build errors - paste relevant lines and add `build failure` label
+  - due to other formulas need revision bumps - suggest to use the following command:
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: /formulas need/formulae needing/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    ```
+    # in this example PR is for `libuv` formula and `urbit` needs revision bump
+    brew bump-revision --message 'for libuv' urbit
+    ```
+- if CI is green and formula `bottle :unneeded` you can merge it through GitHub UI
+- if CI is green and bottles need to be pulled, use: `brew pull --bottle $PR_ID`
+- don't forget to thank the contributor
+  - celebrate the first-time contributors
+- suggest to use `brew bump-formula-pr` next time if this was not a case
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: /a/the/

File: docs/Homebrew-homebrew-core-Merge-Checklist.md [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+  ```
+- is there are test block for other than checking version or printing help? Consider asking to add one
+- if CI failed
+  - due to test block - paste relevant lines and add `test failure` label
+  - due to build errors - paste relevant lines and add `build failure` label
+  - due to other formulas need revision bumps - suggest to use the following command:
+    ```
+    # in this example PR is for `libuv` formula and `urbit` needs revision bump
+    brew bump-revision --message 'for libuv' urbit
+    ```
```
Comment (EricFromCanada) [inline docs/Homebrew-homebrew-core-Merge-Checklist.md]: And note that revision bumps should be separate commits.

**Comment by EricFromCanada** [comment]
Do we only revision-bump dependent formulae if they fail to build in CI? Or do we take a more proactive approach, by e.g. bumping popular dependent formulae for significant upgrades?

**Comment by SMillerDev** [comment]
Only if they fail to build in CI since the tests should fail if a dependency is broken.

**Comment by MikeMcQuaid** [comment]
Looks good enough to merge. Thanks @igas and everyone who reviewed and provided comments. Can iterate on the file now it's in.
